Kudu.Services/Deployment/PushDeploymentController.cs

```diff
@@ -222,6 +237,50 @@ public class PushDeploymentController : ApiController
             repository.Commit(zipDeploymentInfo.Message, zipDeploymentInfo.Author, zipDeploymentInfo.AuthorEmail);
         }
 
+        private async Task LocalZipFetchAndDeploy(IRepository repository, DeploymentInfoBase deploymentInfo, ILogger logger, ITracer tracer)
+        {
+            var zipDeploymentInfo = (ZipDeploymentInfo)deploymentInfo;
+
+            // For this kind of deployment, RepositoryUrl is a local path to the temp zip.
+            var sourceZipFile = zipDeploymentInfo.RepositoryUrl;
+            // the name of the deployed file in D:\home\data\SitePackages\{name}.zip is the 
+            // timestamp in the format yyyMMddHHmmss. 
+            var fileName = $"{DateTime.UtcNow.ToString("yyyyMMddHHmmss")}.zip";
+            var filePath = Path.Combine(_environment.SitePackagesPath, fileName);
+
+            FileSystemHelpers.EnsureDirectory(_environment.SitePackagesPath);
+
+            var info = FileSystemHelpers.FileInfoFromFileName(sourceZipFile);
+            var sizeInMb = (info.Length / (1024f * 1024f)).ToString("0.00", CultureInfo.InvariantCulture);
+
+            var message = String.Format(
+                CultureInfo.InvariantCulture,
+                "Deploying {0} to {1} with size of {2}",
+                info.FullName,
+                filePath,
+                sizeInMb);
+
+            logger.Log(message);
+
+            using (tracer.Step(message))
+            {
+                // If extractTargetDirectory already exists, rename it so we can delete it concurrently with
+                // the unzip (along with any other junk in the folder)
+                var cleanTask = Task.Run(() => DeleteFilesAndDirsExcept(sourceZipFile, null, tracer));
+                var extractTask = Task.Run(() =>
+                {
+                    FileSystemHelpers.CopyFile(sourceZipFile, filePath, overwrite: true);
+                    FileSystemHelpers.WriteAllTextToFile(Path.Combine(_environment.SitePackagesPath, Constants.SiteVersionTxt), fileName);
+                });
+
+                await Task.WhenAll(cleanTask, extractTask);
+            }
+
+            // Needed in order for repository.GetChangeSet() to work.
+            // Similar to what OneDriveHelper and DropBoxHelper do.
+            repository.Commit(zipDeploymentInfo.Message, zipDeploymentInfo.Author, zipDeploymentInfo.AuthorEmail);
```

Comment:
I don't remember how that all works. I thought in typical zipdeploy scenarios there was no repo at all. What is this committing to?

Reply:
You're right. It's not needed here. In regular ZipDeploy while there is no repo, apparently we're creating a NullRepository with 1 commit as above because the rest of the code expects a ChangeSet() for various logs and checks and it was probably easier to add this instead of wiring up all the deployment code to handle no repo scenarios. 

However because we will skip the actual deploy step for run-from-zip, we don't need this here